src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/Macros/GeneratePortNumberConfig.cs

```diff
@@ -0,0 +1,145 @@
+        {
+                    2049, // nfs
+                    3659, // apple-sasl / PasswordServer
+                    4045, // lockd
+                    6000, // X11
+                    6665, // Alternate IRC [Apple addition]
+                    6666, // Alternate IRC [Apple addition]
+                    6667, // Standard IRC [Apple addition]
+                    6668, // Alternate IRC [Apple addition]
+                    6669, // Alternate IRC [Apple addition]
+        };
+
+        private static readonly HashSet<int> AllocatedPorts = new();
+
+        internal GeneratePortNumberConfig(GeneratePortNumberMacro macro, string variableName, string? dataType, int fallback, int low, int high)
+             : base(macro, variableName, dataType)
+        {
+            Port = AllocatePort(low, high, fallback);
+        }
+
+        internal GeneratePortNumberConfig(GeneratePortNumberMacro macro, IGeneratedSymbolConfig generatedSymbolConfig)
+            : base(macro, generatedSymbolConfig.VariableName, generatedSymbolConfig.DataType)
+        {
+            int low = GetOptionalParameterValue(generatedSymbolConfig, nameof(Low), ConvertJTokenToInt, LowPortDefault);
+            int high = GetOptionalParameterValue(generatedSymbolConfig, nameof(High), ConvertJTokenToInt, HighPortDefault);
+            if (low < LowPortDefault)
+            {
+                low = LowPortDefault;
+            }
+
+            if (high > HighPortDefault)
+            {
+                high = HighPortDefault;
+            }
+
+            if (low > high)
+            {
+                low = LowPortDefault;
+                high = HighPortDefault;
+            }
+
+            int fallback = GetOptionalParameterValue(generatedSymbolConfig, "fallback", ConvertJTokenToInt, 0);
+
+            Port = AllocatePort(low, high, fallback);
+        }
+
+        internal int Port { get; }
+
+        internal int Low { get; }
+
+        internal int High { get; }
+
+        private static int AllocatePort(int low, int high, int fallback = 0)
+        {
+            int startPort = CryptoRandom.NextInt(low, high);
+
+            for (int testPort = startPort; testPort <= high; testPort++)
+            {
+                if (TryAllocatePort(testPort, out int port))
+                {
+                    return port;
+                }
+            }
+
+            for (int testPort = low; testPort < startPort; testPort++)
+            {
+                if (TryAllocatePort(testPort, out int port))
+                {
+                    return port;
+                }
+            }
+            return fallback;
+        }
+
+        private static bool TryAllocatePort(int testPort, out int finalPort)
+        {
+            Socket? testSocket = null;
+            finalPort = 0;
+
+            if (UnsafePorts.Contains(testPort))
+            {
+                return false;
+            }
+            lock (LockObj)
```

Comment:
I guess one way to avoid extended locking is to use `ConcurrentBag`:

```
if(unavailablePorts.Contains(port))
{
    return false;
}
// add it to the collection upfront - as it should not be retested in future - whether taken already or whether it's going to be taken now
unavailablePorts.Add(port);
// ... try bind to socket and return true if successfull
```

There is a slight window for concurrency issue (between test and add) - that can be guarded by lock or spinlock - but it's not worth the added complexity of the code, as chance is very low and it will only lead to unnecessary extra check.

Using `ConcurrentBag` will as well allow using just single collection for reserved and taken ports instead of two - so code should simplify even further.

Reply:
What about `ConcurrentDictionary`? Then it eliminates any risk.